src/hash_map.c

```diff
@@ -0,0 +1,562 @@
+    RCUTILS_SET_ERROR_MSG("failed to allocate memory for map data");
+    return ret;
+  }
+
+  hash_map->impl->allocator = *allocator;
+
+  return RCUTILS_RET_OK;
+}
+
+rcutils_ret_t
+rcutils_hash_map_fini(rcutils_hash_map_t * hash_map)
+{
+  HASH_MAP_VALIDATE_HASH_MAP(hash_map);
+  rcutils_ret_t ret = hash_map_deallocate_map(hash_map->impl->map, hash_map->impl->capacity,
+      &hash_map->impl->allocator, true);
+
+  if (RCUTILS_RET_OK == ret) {
+    hash_map->impl->allocator.deallocate(hash_map->impl, hash_map->impl->allocator.state);
+    hash_map->impl = NULL;
+  }
+
+  return ret;
+}
+
+rcutils_ret_t
+rcutils_hash_map_get_capacity(const rcutils_hash_map_t * hash_map, size_t * capacity)
+{
+  HASH_MAP_VALIDATE_HASH_MAP(hash_map);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(capacity);
+  *capacity = hash_map->impl->capacity;
+  return RCUTILS_RET_OK;
+}
+
+rcutils_ret_t
+rcutils_hash_map_get_size(const rcutils_hash_map_t * hash_map, size_t * size)
+{
+  HASH_MAP_VALIDATE_HASH_MAP(hash_map);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(size);
+  *size = hash_map->impl->size;
+  return RCUTILS_RET_OK;
+}
+
+/// Returns true if found or false if it doesn't exist.
+/// key_hash and map_index will always be set correctly
+static bool hash_map_find(
+  const rcutils_hash_map_t * hash_map,   // [in] The hash_map to look up in
+  const void * key,   // [in] The key to lookup
+  size_t * key_hash,   // [out] The key's hashed value
+  size_t * map_index,   // [out] The index into the array of buckets
+  size_t * bucket_index,   // [out] The index of the entry in its bucket
+  rcutils_hash_map_entry_t ** entry)   // [out] Will be set to a pointer to the entry's data
+{
+  size_t bucket_size = 0;
+  rcutils_hash_map_entry_t * bucket_entry = NULL;
+
+  *key_hash = hash_map->impl->key_hashing_func(key);
+  *map_index = (*key_hash) % hash_map->impl->capacity;
+
+  // Find the bucket the entry should be in check that it's valid
+  rcutils_array_list_t * bucket = &(hash_map->impl->map[*map_index]);
+  if (NULL == bucket->impl) {
+    return false;
+  }
+  if (RCUTILS_RET_OK != rcutils_array_list_get_size(bucket, &bucket_size)) {
+    return false;
+  }
+  for (size_t i = 0; i < bucket_size; ++i) {
+    if (RCUTILS_RET_OK != rcutils_array_list_get(bucket, i, &bucket_entry)) {
+      return false;
+    }
+    // Check that the hashes match first as that will be the quicker comparison to quick fail on
+    if (bucket_entry->hashed_key == *key_hash &&
+      (0 == hash_map->impl->key_cmp_func(bucket_entry->key, key)))
+    {
+      *bucket_index = i;
+      *entry = bucket_entry;
+      return true;
+    }
+  }
+
+  return false;
+}
+
+rcutils_ret_t
+rcutils_hash_map_set(rcutils_hash_map_t * hash_map, const void * key, const void * value)
+{
+  HASH_MAP_VALIDATE_HASH_MAP(hash_map);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(key);
+  HASH_MAP_VALIDATE_ARGUMENT_NOT_NULL(value);
+
+  size_t key_hash = 0, map_index = 0, bucket_index = 0;
+  bool already_exists = false;
```

Comment:
@nburek nit: why not declaring it at the time of first use?

Reply:
It's just grouping declarations near the top. They're all used shortly after.